SharkCage/CageManager/CageManager.cpp

```diff
@@ -300,24 +371,93 @@ bool CreateACL(std::unique_ptr<PSID, decltype(local_free_deleter<PSID>)> group_s
 	path_buf.push_back(0);
-	if (::CreateProcess(NULL, path_buf.data(), NULL, NULL, TRUE, 0, NULL, NULL, &info, &process_info))
+
+	if (::CreateProcess(NULL, path_buf.data(), NULL, NULL, TRUE, 0, NULL, NULL, &info, &process_info) != 0)
 	{
 		std::cout << "Failed to start process. Err " << ::GetLastError() << std::endl;
 	}
 
+	std::optional<PROCESS_INFORMATION> process_info_additional_app;
+	if (additional_app.has_value() && additional_app_path.has_value())
+	{
+			std::vector<wchar_t> path_buf_keepass(additional_app_path.value().begin(), additional_app_path.value().end());
+			path_buf_keepass.push_back(0);
+			process_info_additional_app = { 0 };
+			STARTUPINFO info_additional_app = { 0 };
+			info_additional_app.lpDesktop = new_desktop_name_buf.data();
+
+			if (::CreateProcess(NULL, path_buf_keepass.data(), NULL, NULL, TRUE, 0, NULL, NULL, &info_additional_app, &process_info_additional_app.value()))
+			{
+				std::cout << "Failed to start process in lalalalal. Err " << GetLastError() << std::endl;
+			}
+	}
+
+	// Start a thread that will perform the main task of the service
+	//worker_thread = ::CreateThread(NULL, 0, ManagerWorkerThread, NULL, 0, NULL);
+	//::WaitForSingleObject(worker_thread, INFINITE);
+
 	desktop_thread.join();
+	receiving_messages = false;
 
 	//::SendMessage(nullptr /*main handle of new process*/, WM_CLOSE, NULL);
 	// with timeout, if nothing happens, terminate process
+	if (process_info_additional_app.has_value() && !::TerminateProcess(process_info_additional_app.value().hProcess, 0))
+	{
+		std::cout << "Failed to terminate process Err " << ::GetLastError() << std::endl;
+	}
+
 	if (!::TerminateProcess(process_info.hProcess, 0))
 	{
 		std::cout << "Failed to terminate process Err " << ::GetLastError() << std::endl;
 	}
 
 	// wait for the process to exit
 	::WaitForSingleObject(process_info.hProcess, INFINITE);
+	
+	if (process_info_additional_app.has_value())
+	{
+		::CloseHandle(process_info_additional_app.value().hProcess);
+		::CloseHandle(process_info_additional_app.value().hThread);
+	}
+
 	::CloseHandle(process_info.hProcess);
 	::CloseHandle(process_info.hThread);
-	
+
 	return true;
 }
+
+DWORD WINAPI ManagerWorkerThread(LPVOID)
+{
+	// periodically check if the service has been requested to stop
+	// But does not work properly because netwirkMgr.listen() is a blocking call.
+	while (receiving_messages)
+	{
+		// Look for messages and parse them
+		auto message = network_manager.Listen();
+		auto message_result = WaitForMessage(message);
+
+		if (message_result.has_value())
+		{
+			if (BeginsWith(message_result.value(), ManagerMessageToString(ManagerMessage::RESTART_ADDITIONAL_APP)))
+			{
+				std::wstring new_desktop_name = L"shark_cage_desktop";
+				std::vector<wchar_t> new_desktop_name_buf(new_desktop_name.begin(), new_desktop_name.end());
+				new_desktop_name_buf.push_back(0);
+
+				std::wstring path_keepass = L"C:\\Program Files (x86)\\KeePass Password Safe 2\\KeePass.exe";
```

Comment:
* hard coded path
* this code block looks like it shares a lot of code with the normal start of the processes
* why do you only check for the additional app restart and not the normal app restart?

Reply:
This was test code for the restarting part of the application, is complete removed now.